examples/usage.rs

```diff
@@ -0,0 +1,58 @@
+#![feature(try_blocks)]
+// Overrides the std library's `Result` type.
+
+use std::fs::File;
+use std::io;
+
+enum MyError {
+    Unlucky,
+    Io(io::Error),
+    TooSmall(u64),
+}
+
+impl From<io::Error> for MyError {
+    fn from(e: io::Error) -> Self {
+        Self::Io(e)
+    }
+}
+
+fn file_size(path: &str) -> propagate::Result<u64, MyError> {
+    try {
+        // The `?` operator coerces `std::result::Result<_, io::Error>` into `Result<_, MyError>`.
+        let size = File::open(path)?.metadata()?.len();
+
+        if size < 1024 {
+            Err(MyError::TooSmall(size))?
+        }
+
+        size
+    }
+}
+
+fn maybe_file_size(path: &str) -> propagate::Result<u64, MyError> {
+    let lucky = (path.len() % 2) == 0;
+
+    try {
+        if !lucky {
+            Err(MyError::Unlucky)?
+        }
+
+        file_size(path)?
+    }
+}
+
+fn main() {
+    let result = maybe_file_size("foo.txt");
+    match result {
+        propagate::Ok(size) => println!("File size: {} KiB", size / 1024),
+        propagate::Err(err) => {
+            // Dereference `err` to get the actual error value.
+            match &*err {
+                MyError::Unlucky => println!("Not this time!"),
+                MyError::Io(e) => println!("I/O error: {}", e),
+                MyError::TooSmall(size) => println!("File too small: {} bytes", size),
+            }
+            println!("Backtrace: {:?}", err.stack());
+        }
+    }
```

Comment:
I like the usage of Deref to get at the inner error type here though I don't love how we can't get the deref coersion to happen implicitly here :/

I want to investigate alternative ways to expose the internal error and the stack to see if we might be able to do this more intuitively but as of yet I can't think of anything better than what you came up with.

Reply:
Been thinking about this a bit and my expectation is that it might be that manually matching on `propagate::Result` is rarely necessary. My hope is that most of the time where you need to react to a specific error it will be before coercing it to a `propagate::Result`, and you'll only be tracing errors that you need to report to users, and that in this case we can focus on making it easy to report an error and its trace so you never have to manually write out a match like this.